### ISSUE TITLE
Could NOT find X11 (missing: X11_X11_LIB)

### ISSUE THREAD

**Issue by wangshucheng** [glfw/glfw#1957]
wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ sudo apt-get install libx11-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
libx11-dev is already the newest version (2:1.6.9-2ubuntu1.2).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.

wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ cmake ..
-- Using X11 for window creation
CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find X11 (missing: X11_X11_LIB)
Call Stack (most recent call first):
  /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.16/Modules/FindX11.cmake:366 (find_package_handle_standard_args)
  CMakeLists.txt:203 (find_package)

-- Configuring incomplete, errors occurred!
See also "/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build/CMakeFiles/CMakeOutput.log".
See also "/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build/CMakeFiles/CMakeError.log".

wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ sudo apt-cache search x11-dev
libx11-dev - X11 client-side library (development headers)
libxkbcommon-x11-dev - library to create keymaps with the XKB X11 protocol - development files
libghc-gi-gdkx11-dev - GDKX11 bindings
libghc-x11-dev - Haskell X11 binding for GHC
libnx-x11-dev - nxagent's libNX_X11 client-part library (development headers)
librust-cairo-sys-rs+x11-dev - FFI bindings to libcairo - feature "x11"
librust-x11-dev - X11 library bindings for Rust - Rust source code

**Comment by elmindreda** [comment]
I'm not sure exactly what is causing the error above but you probably want to install the full `xorg-dev` package if you haven't already, not just `libx11-dev`. GLFW needs the headers from several other packages brought in by `xorg-dev` in order to build.

If the problem remains with that package installed then that's definitely a bug somewhere.

**Comment by wangshucheng** [comment]
> I'm not sure exactly what is causing the error above but you probably want to install the full `xorg-dev` package if you haven't already, not just `libx11-dev`. GLFW needs the headers from several other packages brought in by `xorg-dev` in order to build.
> 
> If the problem remains with that package installed then that's definitely a bug somewhere.

root@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build# sudo apt-get install xorg-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
xorg-dev is already the newest version (1:7.7+19ubuntu14).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.


Thank you for your answer. But I think xorg-dev is also completely installed.

**Comment by elmindreda** [comment]
You're running Ubuntu on WSL?

I don't think I've ever seen CMake fail to find X11 installed from the regular packages on a Debian derived system so I really don't know what could be causing this.

**Comment by Velaseriat** [comment]
oof. I'm on the same boat with WSL2. I guess it's a WSL thing.

**Comment by elmindreda** [comment]
It could be a WSL thing but also I use WSL daily for building and testing the X11 backend and have not seen this issue.

**Comment by LeonFretter** [comment]
For me this happened when trying to compile with clang. With gcc it worked fine. 
Can anyone maybe try reproducing with clang?

**Comment by Seratna** [comment]
I used cmake 3.22.3 and ran into the same problem. I switched back to cmake 3.21.6 and every thing is OK.

**Comment by jinfagang** [comment]
Damn it. Got same errro here. I have installed libxorg-dev libx11-dev everything it needed, but glfw just can not found it.

Any solution? Mine cmake 3.22.3

**Comment by jinfagang** [comment]
btw, downgrade or upgrade cmake doesn't help at all.

**Comment by elmindreda** [comment]
I don't know why CMake is failing to find `libX11.so` on some systems. That should possibly be reported to CMake.

In the meantime, if you know where the library is on your system you should be able to work around this issue with [CMAKE_LIBRARY_PATH](https://cmake.org/cmake/help/latest/variable/CMAKE_LIBRARY_PATH.html) by adding `-DCMAKE_LIBRARY_PATH=/path/to/lib` on the CMake command-line.

**Comment by jinfagang** [comment]
For anyone got this error, recommend using `cmake -DGLFW_USE_WAYLAND=ON ..  ` and it works

works like a charm. And what even more strangely, if using this, it will found x11 again......

**Comment by ben-clayton** [comment]
This issue appears to be reproducible with Clang 14, but not with Clang 12 (!?)

**Comment by jinfagang** [comment]
Damn it! Now  `cmake -DGLFW_USE_WAYLAND=ON .. ` doesn't work on same machine....... same env................. Just another time...............

I am really doubt this logic ,this X11 thing, it's just like a strange shit.

**Comment by jinfagang** [comment]
OK, I found the problem, the f**k guilty guy behind it was **anaconda**.....

this guy just hide all lib from my system.... Be note even if you deactivated it, it still effect your system.

**Comment by binbinJiaYou** [comment]
> wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ sudo apt-get install libx11-dev Reading package lists... Done Building dependency tree Reading state information... Done libx11-dev is already the newest version (2:1.6.9-2ubuntu1.2). 0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
> 
> wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ cmake .. -- Using X11 for window creation CMake Error at /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message): Could NOT find X11 (missing: X11_X11_LIB) Call Stack (most recent call first): /usr/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE) /usr/share/cmake-3.16/Modules/FindX11.cmake:366 (find_package_handle_standard_args) CMakeLists.txt:203 (find_package)
> 
> -- Configuring incomplete, errors occurred! See also "/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build/CMakeFiles/CMakeOutput.log". See also "/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build/CMakeFiles/CMakeError.log".
> 
> wangshucheng@wangshucheng:/mnt/c/Users/wangsc/Downloads/glfw/glfw-3.3.4/build$ sudo apt-cache search x11-dev libx11-dev - X11 client-side library (development headers) libxkbcommon-x11-dev - library to create keymaps with the XKB X11 protocol - development files libghc-gi-gdkx11-dev - GDKX11 bindings libghc-x11-dev - Haskell X11 binding for GHC libnx-x11-dev - nxagent's libNX_X11 client-part library (development headers) librust-cairo-sys-rs+x11-dev - FFI bindings to libcairo - feature "x11" librust-x11-dev - X11 library bindings for Rust - Rust source code

你可以试一下这个：
https://blog.csdn.net/qq_39845665/article/details/125462900

**Comment by ben-clayton** [comment]
TL;DR: My issues went away after updating to CMake to [`v3.24.1`](https://github.com/Kitware/CMake/releases/tag/v3.24.1).

If that still doesn't work, consider setting [`CMAKE_LIBRARY_PATH`](https://cmake.org/cmake/help/latest/variable/CMAKE_LIBRARY_PATH.html) to include the path to X11 in your application's `CMakeLists.txt`.

---

I spent some time investigating why `clang-12` works and `clang-14` doesn't. These are my findings:

`X11_X11_LIB` is set in [`FindX11.cmake`](https://github.com/Kitware/CMake/blob/d69d0316da2bd8ead5f9ebd03180e329a6180a1d/Modules/FindX11.cmake#L182), by using CMake's [`find_library`](https://cmake.org/cmake/help/latest/command/find_library.html) command. `FindX11.cmake` has [a hardcoded list of search paths](https://github.com/Kitware/CMake/blob/d69d0316da2bd8ead5f9ebd03180e329a6180a1d/Modules/FindX11.cmake#L112-L118), however on my system (ubuntu 21.04) `libX11.so` was found at `/usr/lib/x86_64-linux-gnu/libX11.so`, which isn't part of this list.

[The documentation for `find_library()`](https://cmake.org/cmake/help/latest/command/find_library.html) states that it falls back to searching at `<prefix>/lib/<arch>` if [`CMAKE_LIBRARY_ARCHITECTURE`](https://cmake.org/cmake/help/latest/variable/CMAKE_LIBRARY_ARCHITECTURE.html#variable:CMAKE_LIBRARY_ARCHITECTURE) is set.

Low and behold, `CMAKE_LIBRARY_ARCHITECTURE` is different between clang-12 and clang-14:
* `clang-12`:  `CMAKE_LIBRARY_ARCHITECTURE=x86_64-linux-gnu`
* `clang-14`:  `CMAKE_LIBRARY_ARCHITECTURE=x86_64-unknown-linux-gnu`

I tried digging further into where `CMAKE_LIBRARY_ARCHITECTURE` was set, and quickly found that this had [all been changed](https://github.com/Kitware/CMake/commit/657fc3a9a719831f9943c552e86e58ae283dad43) since my build of CMake. I grabbed a newer build, and magically the X11 issue is now gone.

**Comment by mikejsavage** [comment]
If you're building from glfw master you don't need libx11.so present anymore. This still fails but you can just do `cmake -DX11_X11_LIB=who_cares"` as a workaround

**Comment by elmindreda** [comment]
With the findings of @ben-clayton above, I believe this can be closed as a resolved external bug.

**Comment by xuyaojian123** [comment]
`cmake -DGLFW_USE_WAYLAND=OFF .. ` It seems compiled successfully.

**Comment by MrGuyman99** [comment]
Hey! I know this issue is closed, but if anyone is using a steam deck to develop thing (Since I'm traveling I've been doing that) I actually ran into this issue again. None of the above fixes worked, although forcibly reinstalling all the X11 libraries actually fixed the issue! Probably no one will see this, but I'd be glad to help is someone did
